homeassistant/components/upcloud/__init__.py

```diff
@@ -52,65 +64,200 @@
             {
                 vol.Required(CONF_USERNAME): cv.string,
                 vol.Required(CONF_PASSWORD): cv.string,
-                vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL): cv.time_period,
+                vol.Optional(
+                    CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+                ): cv.time_period,
             }
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
-    """Set up the UpCloud component."""
-    conf = config[DOMAIN]
-    username = conf.get(CONF_USERNAME)
-    password = conf.get(CONF_PASSWORD)
-    scan_interval = conf.get(CONF_SCAN_INTERVAL)
+class UpCloudDataUpdateCoordinator(
+    DataUpdateCoordinator[Dict[str, upcloud_api.Server]]
+):
+    """UpCloud data update coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistantType,
+        *,
+        cloud_manager: upcloud_api.CloudManager,
+        update_interval: timedelta,
+        username: str,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=f"{username}@UpCloud", update_interval=update_interval
+        )
+        self.cloud_manager = cloud_manager
+        self.unsub_handlers: List[CALLBACK_TYPE] = []
+
+    async def async_update_config(self, config_entry: ConfigEntry) -> None:
+        """Handle config update."""
+        self.update_interval = timedelta(
+            seconds=config_entry.options[CONF_SCAN_INTERVAL]
+        )
+
+    async def _async_update_data(self) -> Dict[str, upcloud_api.Server]:
+        return {
+            x.uuid: x
+            for x in await self.hass.async_add_executor_job(
+                self.cloud_manager.get_servers
+            )
+        }
+
+
+@dataclasses.dataclass
+class UpCloudHassData:
+    """Home Assistant UpCloud runtime data."""
+
+    coordinators: Dict[str, UpCloudDataUpdateCoordinator] = dataclasses.field(
+        default_factory=dict
+    )
+    scan_interval_migrations: Dict[str, int] = dataclasses.field(default_factory=dict)
+
+
+async def async_setup(hass: HomeAssistantType, config) -> bool:
+    """Set up UpCloud component."""
+    domain_config = config.get(DOMAIN)
+    if domain_config:
+        _LOGGER.warning(
+            "Loading upcloud via top level config is deprecated and no longer "
+            "necessary as of 0.116. Please remove it from your YAML configuration."
+        )
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: domain_config[CONF_USERNAME],
+                    CONF_PASSWORD: domain_config[CONF_PASSWORD],
+                },
+            )
+        )
+
+        if domain_config[CONF_SCAN_INTERVAL]:
+            hass.data.setdefault(
```

Comment:
No need to use setdefault, `async_setup` is only ever run once.

Reply:
Changed.